src/main/scala/viper/gobra/frontend/info/implementation/resolution/NameResolution.scala

```diff
@@ -301,9 +309,56 @@ trait NameResolution { this: TypeInfoImpl =>
           findField(litType, id).getOrElse(UnknownEntity())
         } else symbTableLookup(n) // otherwise it is just a variable
 
+      case tree.parent.pair(id: PIdnUse, tree.parent.pair(_: PIdentifierKey, tree.parent(spec: PClosureSpecInstance))) =>
+        lookupFunctionMemberOrLit(spec.func).flatMap(func => lookupParamForClosureSpec(id, func)) match {
+          case Some(p: InParameter) => p
+          case _ => UnknownEntity() // only in-parameter names can be used as keys
+        }
+
+      case n: PIdnUse if tryEnclosingClosureImplementationProof(n).nonEmpty =>
+        val proof = tryEnclosingClosureImplementationProof(n).get
+        lookupFunctionMemberOrLit(proof.impl.spec.func).flatMap(func => lookupParamForClosureSpec(n, func)) match {
+          case Some(p: InParameter) => p
+          case Some(p: OutParameter) => p
+          case _ => symbTableLookup(n)
+        }
+
       case n => symbTableLookup(n)
     }
 
+  private def lookupParamForClosureSpec(id: PIdnUse, func: ActualDataEntity with WithArguments with WithResult): Option[ActualVariable] = {
+    def namedParam(p: PParameter): Option[PNamedParameter] = p match {
+      case p: PNamedParameter => Some(p)
+      case PExplicitGhostParameter(p: PNamedParameter) => Some(p)
+      case _ => None
+    }
+
+    // Within a spec implementation proof or closure instance, consider all arguments of the spec non-ghost and all results ghost.
+    // This is to be as permissive as possible at this point, since all ghostness-related checks are done later on.
+    lazy val inParam = func.args.map(namedParam).find(p => p.nonEmpty && p.get.id.name == id.name)
+      .map(p => InParameter(p.get, ghost = false, addressable = false, func.context))
+    if (inParam.nonEmpty) Some(inParam.get)
+    else  func.result.outs.map(namedParam).find(p => p.nonEmpty && p.get.id.name == id.name)
+      .map(p => OutParameter(p.get, ghost = true, addressable = false, func.context))
+  }
```

Comment:
could this be simplified to the following?
```scala
    lazy val inParam = func.args.flatMap(namedParam)
      .find(_.id.name == id.name)
      .map(InParameter(_, ghost = false, addressable = false, func.context))
    
    inParam.orElse{
      func.result.outs.flatMap(namedParam)
        .find(_.id.name == id.name)
        .map(OutParameter(_, ghost = true, addressable = false, func.context))
    }
  }
```

Reply:
Also, why did you make `inParam` lazy? It seems that the value needs to be computed anyway right after the declaration, so lazy here may not be needed

Reply:
Yes, good suggestion

Reply:
I'm not sure why I made `inParam` lazy, I probably implemented this in a slightly different way at the beginning which made it being lazy useful, and when I changed it I didn't notice it and forgot to remove it

Reply:
Done